agent/consul/gateways/gateway_meta.go

```diff
@@ -0,0 +1,143 @@
+package gateways
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// GatewayMeta embeds both a BoundAPIGateway and its corresponding APIGateway.
+type GatewayMeta struct {
+	// Bound is the bound config entry for a given gateway.
+	Bound *structs.BoundAPIGatewayConfigEntry
+	// Gateway is the APIGateway config entry for the gateway.
+	Gateway *structs.APIGatewayConfigEntry
+}
+
+// GetGatewayMeta queries the state store for an API Gateway and a Bound API
+// Gateway matching the given name and enterprise meta and returns a GatewayMeta
+// struct containing both.
+func GetGatewayMeta(store *state.Store, name string, entMeta *acl.EnterpriseMeta) (*GatewayMeta, error) {
+	_, bound, err := store.ConfigEntry(nil, structs.BoundAPIGateway, name, entMeta)
```

Comment:
~~Is it possible that `bound` is `nil` here? If so, is that okay?~~

Edit: Seeing below that they can be `nil` and we handle that in various places